src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,509 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from pyspark.sql import functions as F
+
+from mozanalysis.utils import add_days
+
+
+class Experiment(object):
+    """Get DataFrames of experiment data; store experiment metadata.
+
+    The methods here query data in a way compatible with the following
+    principles, which are important for experiment analysis:
+
+    - The population of clients in each branch must have the same
+        properties, aside from the intervention itself and its
+        consequences; i.e. there must be no underlying bias in the
+        branch populations.
+    - Every client should have an equal opportunity to convert, so that
+        we are measuring the same thing for each client.
+
+    So that our analyses follow these abstract principles, we follow
+    these rules:
+
+    - Start with a list of all clients who enrolled.
+    - We can filter this list of clients only based on information known
+        to us at or before the time that they enrolled, because later
+        information might be causally connected to the intervention.
+    - For any given metric, every client gets a non-null value; we don't
+        implicitly ignore anyone, even if they churned and stopped
+        sending data.
+    - Typically if an enrolled user no longer qualifies for enrollment,
+        we'll still want to include their data in the analysis, unless
+        we're explicitly using stats methods that handle censored data.
+
+
+    Example usage:
+        cd = spark.table('clients_daily')
+
+        experiment = Experiment(
+            experiment_slug='pref-flip-defaultoncookierestrictions-1506704',
+            start_date='20181215',
```

Comment:
IIUC the `experiments` table filters `clients_daily` based on the contents of the `experiments` map. At present, if a client does something that means they no longer match the normandy recipe, then they unenroll from the experiment and their future data won't appear in the `experiments` table - so they will appear to churn. This process is often branch-dependent and therefore is a source of bias. I can point you to a couple recent experiments this happened to (both of which were high volume so didn't appear in the `experiments` table anyway).

Reply:
I understand the unenrollment issue, but would that be an issue here? We're looking for the first date we saw data for any client, correct? If so, `experiments` should still work. IIUC, `experiments` has a row for each ping noting an active experiment.

Reply:
Some experiments don't appear in `experiments` - e.g. high volume experiments, and I believe non-desktop experiments.

Also if we incorrectly filter dates so as to ignore the first 2 days of the experiment, then if we generate `enrollments` from enrollment events then we'll lose two days of enrolled users, but if we generate `enrollments` from `experiments` then a non-churned self-selected subset of these users will be included in `enrollments` with an incorrect enrollment date. I'd rather fail in the former fashion because the impact is more obvious and doesn't compromise the experiment by introducing potential bias.

Reply:
Ah! I didn't know about the experiment `events` table! That makes a lot of sense.

Part of the question still holds - why not query `events` directly to identify the start date. But, I believe you effectively answered this with the `get_per_client_data` discussion. i.e. we want flexibility here and we can always have stronger defaults in mozreport. 

Thanks for walking me through this!

Reply:
> why not query `events` directly to identify the start date
Yes, this would be safer, but there is one final chicken that needs to come before that egg: when querying `events` we should specify a minimum date so that we can efficiently filter out prehistoric events.

So the current need to supply the "maximum lower bound" on relevant data could be relaxed to only require a lower bound, which would be a bit safer. Is that benefit worth losing this method's laziness? *shrugs*

Reply:
Strong point. I agree, the current implementation makes a lot more sense.